doc: Add study guide for the things security

pradeeka7 opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Add study guide for the Things Certified Security exam.

#### Screenshots
<!-- Post a screenshot of your rendered content
NOTE: This section is optional.
-->

![image](https://user-images.githubusercontent.com/36432168/149671380-85c990cf-f3a6-4a16-88f7-522f30eb5c57.png)


#### Changes
<!-- What are the changes made in this pull request? -->

- ...
- ...

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@benolayinka please review

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Run Locally: Verified that the docs build using `make server`, posted screenshots, verified external links.
- [X] Style Guidelines: Documentation obeys style guidelines in [CONTRIBUTING](CONTRIBUTING.md).
- [X] Commits: Commit messages follow guidelines in [CONTRIBUTING](CONTRIBUTING.md), there are no fixup commits left.

pradeeka7 committed:
doc: Fix study guide for the things security

benolayinka commented:
@nejraselimovic can you review this?

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+---
+title: "The Things Certified Security"
```
```suggestion
title: "Preparing for The Things Certified Security"
```

nejraselimovic replied:
In general, it would be good to keep things as consistent as possible to what we already have for The Things Certified Fundamentals -> https://www.thethingsnetwork.org/docs/lorawan/preparing-certification/

nejraselimovic replied:
Also, I think there should be some cross-linking between these two pages. We mention The Things Certified Security in Preparing for The Things Certified Fundamentals page, but there's no link to this study material

pradeeka7 replied:
https://www.thethingsnetwork.org/docs/lorawan/preparing-certification/- This page presents how to prepare for the Things Fundamental exam. @benolayinka Can you move the 'Preparing for The Things Certified Fundamentals' page to the 'THE THINGS FUNDAMENTALS' section?
![image](https://user-images.githubusercontent.com/36432168/150295292-395a8e0c-428c-4153-877d-3e74a31985e7.png)

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+section: Additional Information
+---
+
+This study guide has been written for students who are planning to take **The Things Certified Security** Exam. It explains the following fundamental properties in detail that supported LoRaWAN security:
```
```suggestion
This study guide has been written for students who are planning to take **The Things Certified Security** exam. It explains the following fundamental properties that support LoRaWAN security in detail:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- Replay protection
+- Eavesdropping
+- Spoofing
+- Securing communication between network elements (backend interfaces).
```
```suggestion
In addition to that, we briefly discuss the following security mechanisms:

- Replay protection
- Eavesdropping
- Spoofing
- Securing communication between network elements (backend interfaces)
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- Spoofing
+- Securing communication between network elements (backend interfaces).
+
+We hope that you find the content in this study material useful and that the knowledge you acquire will help you to achieve the passing score you want in the exam.
```
```suggestion
We hope that you will find this study material useful and that the knowledge you acquire will help you pass The Things Certified Security exam.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+## Understanding Message Types
+
+LoRaWAN networks use several MAC message types for different purposes:
```
```suggestion
LoRaWAN networks use several MAC message types for different purposes.

```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+## Understanding Message Types
+
+LoRaWAN networks use several MAC message types for different purposes:
+There are three MAC messages used for the device activation/registration.
```
```suggestion
There are three MAC messages used for the device activation/registration:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- Join-accept
+- Rejoin-request
+ 
+There are four MAC messages used to carry MAC commands and application data together or separately. These messages are known as **data messages**.
```
```suggestion
There are four MAC messages used to carry MAC commands and application data (together or separately), known as **data messages**:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- Confirmed data uplink
+- Confirmed data downlink
+ 
+It is worth knowing the MAC message format before learning how the LoRaWAN security is implemented. The following figure shows the breakdown of a MAC message.
```
```suggestion
It is worth knowing the MAC message format before learning how the LoRaWAN security is implemented. A MAC message format breakdown is shown on the image below.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+It is worth knowing the MAC message format before learning how the LoRaWAN security is implemented. The following figure shows the breakdown of a MAC message.
+
+{{< figure src="mac-message-breakdown.png" alt="MAC Message breakdown" >}}
+*Image credit: LoRaWAN 1.1 Specification document*
```
```suggestion
{{< figure src="mac-message-breakdown.png" alt="MAC Message breakdown" caption="Image credit: LoRaWAN Specification v1.1" >}}
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+{{< figure src="mac-message-breakdown.png" alt="MAC Message breakdown" >}}
+*Image credit: LoRaWAN 1.1 Specification document*
+
+MAC messages carry a payload. The messages used for the **join procedure** have the following payloads. These payloads consist of several fields.
```
```suggestion
MAC messages carry a payload. The messages used for the join procedure have the following payloads, which consist of several fields:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- Rejoin-request
+ 
+The payload carried by the **data message** is called **MACPayload** (also called MAC payload). The term Payload refers to the actual data in a packet. *E.g. application data, MAC commands*. Headers are appended to the payload for transport and then discarded at their destination.
+The MACPayload consists of **FHDR** (Frame Header), **FPort** (Port Field), and the **FRMPayload** (Frame Payload) fields. Only the **FRMPayload** field and **FOpts** field can carry information. The **PHYPayload** is created by adding **MHDR** (MAC Header) and **MIC** (Message Integrity Code) to the **MACPayload**.
```
```suggestion
The payload carried by the data message is called **MACPayload** (MAC Payload). The term payload refers to the actual data in a packet, e.g. application data and/or MAC commands. Headers are appended to the payload for transport and then discarded at the message destination.
The MACPayload consists of **FHDR** (Frame Header), **FPort** (Port Field), and the **FRMPayload** (Frame Payload) fields. Only the **FRMPayload** field and **FOpts** field can carry information. The **PHYPayload** is created by adding **MHDR** (MAC Header) and **MIC** (Message Integrity Code) to the **MACPayload**.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+The payload carried by the **data message** is called **MACPayload** (also called MAC payload). The term Payload refers to the actual data in a packet. *E.g. application data, MAC commands*. Headers are appended to the payload for transport and then discarded at their destination.
+The MACPayload consists of **FHDR** (Frame Header), **FPort** (Port Field), and the **FRMPayload** (Frame Payload) fields. Only the **FRMPayload** field and **FOpts** field can carry information. The **PHYPayload** is created by adding **MHDR** (MAC Header) and **MIC** (Message Integrity Code) to the **MACPayload**.
+
+## Join Procedure: How mutual authentication is done?
```
```suggestion
## Join Procedure
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+LoRaWAN provides you with two options to activate an end device with a LoRaWAN network.
+
+- ABP (Activation-by-Personalisation) - not secured
+- OTAA (Over-The-Air-Activation) - secured and recommended
```
```suggestion
LoRaWAN provides two end device activation methods:

- ABP (Activation-by-Personalisation) - has some security vulnerabilities
- OTAA (Over-The-Air-Activation) - highly recommended
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- ABP (Activation-by-Personalisation) - not secured
+- OTAA (Over-The-Air-Activation) - secured and recommended
+ 
+Out of them, **OTAA** is the most secure way to register/activate an end device on a network. It ensures that only authorised end devices can access the LoRaWAN network by proving that both the end device and the network have the knowledge of **root keys** which is known as **mutual authentication**.
```
```suggestion
OTAA is the most secure way to activate an end device on a network. It ensures that only authorized end devices can access the LoRaWAN network by proving that both the end device and the network have the knowledge of **root keys**. This concept is also known as **mutual authentication**.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+Out of them, **OTAA** is the most secure way to register/activate an end device on a network. It ensures that only authorised end devices can access the LoRaWAN network by proving that both the end device and the network have the knowledge of **root keys** which is known as **mutual authentication**.
+ 
+{{< note "Root keys are unique 128-bit Advanced Encryption Standard (AES) symmetric keys." />}}
```
```suggestion
Root keys used in this procedure are unique 128-bit AES (Advanced Encryption Standard) symmetric keys.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+{{< note "Root keys are unique 128-bit Advanced Encryption Standard (AES) symmetric keys." />}}
+ 
+The following **keys** and **identifiers** must be stored in an end device that is intended to use OTAA as the activation method.
```
```suggestion
The following keys and identifiers must be stored in an OTAA end device:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+The following **keys** and **identifiers** must be stored in an end device that is intended to use OTAA as the activation method.
+
+- **Root keys** - **LoRaWAN v1.0.x** uses a single root key called **AppKey**. **LoRaWAN v1.1** uses two root keys called **AppKey** and **NwkKey**)
```
```suggestion
- **Root keys** - LoRaWAN v1.0.x uses a single root key called **AppKey**. LoRaWAN v1.1 uses two root keys called **AppKey** and **NwkKey**.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+The following **keys** and **identifiers** must be stored in an end device that is intended to use OTAA as the activation method.
+
+- **Root keys** - **LoRaWAN v1.0.x** uses a single root key called **AppKey**. **LoRaWAN v1.1** uses two root keys called **AppKey** and **NwkKey**)
+- **DevEUI** - As a best practice, the DevEUI should also be available on a device label, for device administration. Root key/s for a given end device are identified by its DevEUI.
```
```suggestion
- **DevEUI** - a globally unique 64-bit end device identifier, assigned by the device manufacturer or device owner. Root key/s for a given end device are identified by its DevEUI.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- **LoRaWAN v1.0.x** - AppKey is stored in the Network Server
+- **LoRaWAN v1.1** - AppKey and NwkKey are stored in the Join Server.
+ 
+The storing process of **root keys** and **ids** in both the end device and the network (Network Server or the Join Server) is known as **provisioning**.
```
```suggestion
The matching root keys should be stored on the network side. In LoRaWAN v1.0.x AppKey is stored in the Network Server, while in LoRaWAN v1.1 AppKey and NwkKey are stored in the Join Server. The process of storing root keys and identifiers on both ends is known as **provisioning**.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+- JoinEUI
+- DevEUI 
+- DevNonce
```
```suggestion
The join procedure is started by sending a **Join-request** message by an end device. The Join-request message consists of the following fields:

- AppEUI/JoinEUI
- DevEUI 
- DevNonce
 
Note that LoRaWAN v1.1 and v1.0.4 use term JoinEUI instead of AppEUI.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- DevEUI 
+- DevNonce
+ 
+The Join-request frame **IS NOT encrypted** (sent in clear text) but **integrity protected** by computing the **MIC** (Message Integrity Code). The MIC is computed over all the fields in the message using **[AES-CMAC](https://datatracker.ietf.org/doc/html/rfc4493)** (AES Cipher-based Message Authentication Code) **algorithm** with the **AppKey** (v1.0.x) or **NwkKey** (v1.1). **The computed MIC i ...
```
```suggestion
The Join-request message is not encrypted, i.e. it is sent in plain text. However, it has its **integrity protected** by computing the **MIC** (Message Integrity Code). The MIC is computed over all message fields using **[AES-CMAC](https://datatracker.ietf.org/doc/html/rfc4493)** (AES Cipher-based Message Authentication Code) algorithm together with the root keys. The computed MIC is then added to the message itself. See images explaining this process below:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+The Join-request frame **IS NOT encrypted** (sent in clear text) but **integrity protected** by computing the **MIC** (Message Integrity Code). The MIC is computed over all the fields in the message using **[AES-CMAC](https://datatracker.ietf.org/doc/html/rfc4493)** (AES Cipher-based Message Authentication Code) **algorithm** with the **AppKey** (v1.0.x) or **NwkKey** (v1.1). **The computed MIC i ...
+
+**v1.0.x**
```
```suggestion
**LoRaWAN v1.0.x**
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+**v1.0.x**
+{{< figure src="join-request-mic-10x.png" alt="Join-request MIC in v1.0.x" >}}
+
+**v1.1**
```
```suggestion
**LoRaWAN v1.1**
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+**v1.1**
+{{< figure src="join-request-mic-11.png" alt="Join-request MIC in v1.1" >}}
+
+The **MIC** enables the network to check if the **AppKey (1.0.x)/NwkKey (1.1)** is correct.
```
```suggestion
The MIC enables the network to check if the AppKey (1.0.x) and NwkKey (1.1) are correct.
```

nejraselimovic replied:
move this up, put it after this sentence "However, it has its **integrity protected** by computing the **MIC** (Message Integrity Code)."

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+The **MIC** enables the network to check if the **AppKey (1.0.x)/NwkKey (1.1)** is correct.
+
+The following two sub sections describe how the network process the Join-request message.
```
```suggestion
The following two subsections describe how the network processes the Join-request message.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+### In LoRaWAN v1.0.x
+
+After the **Join-request** frame receives, the **Network Server** (assume that the Network Server and the Application Server are co-located) computes the **MIC** over all the fields in the message (except the MIC field) using **AES-CMAC** algorithm with the **AppKey**. 
```
```suggestion
After the Join-request frame is received, the Network Server (assume that the Network Server and the Application Server are co-located) computes the MIC over all message fields (except the MIC field) using AES-CMAC algorithm with the AppKey. 
```

nejraselimovic replied:
There's really no need to constantly highlight these terms, they are well explained in the text above. Please do that only when you want to introduce a new term or highlight something really important. It's practically against our documentation guidelines.

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+After the **Join-request** frame receives, the **Network Server** (assume that the Network Server and the Application Server are co-located) computes the **MIC** over all the fields in the message (except the MIC field) using **AES-CMAC** algorithm with the **AppKey**. 
+ 
+Then the computed MIC is compared with the MIC field in the Join-request frame. If both MICs are matched, the **Network Server** will **authenticate** the Join-request frame, otherwise discards the Join-request message. This ensures that the Join-request message was not altered by any **malicious actor** during its journey.
```
```suggestion
Then the computed MIC is compared to the MIC field in the Join-request frame. If these MICs are matched, the Network Server will authenticate the Join-request frame, otherwise the Join-request message gets discarded. This ensures that the Join-request message was not altered by any malicious actor during its transmission.
```

pradeeka7 replied:
keep this word: 'journey'

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+Then the computed MIC is compared with the MIC field in the Join-request frame. If both MICs are matched, the **Network Server** will **authenticate** the Join-request frame, otherwise discards the Join-request message. This ensures that the Join-request message was not altered by any **malicious actor** during its journey.
+ 
+If the end device is allowed to join the network and the MIC is correct, the Network Server prepaires a **Join-accept** message. The Join-accept message consists of the following fields.
```
```suggestion
If the end device is allowed to join the network and the MIC is correct, the Network Server prepaires a **Join-accept** message. The Join-accept message consists of the following fields:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- **RxDelay** - 1 byte, delay between TX and RX
+- **CFList** - 16 bytes, optional list of channel frequencies
+ 
+The **MIC** is calculated over the following fields in the **Join-accept** message using the **AES-CMAC** algorithm with the **AppKey**. 
```
```suggestion
The MIC is calculated over the following Join-accept message fields using the AES-CMAC algorithm with the AppKey: 
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+The **MIC** is calculated over the following fields in the **Join-accept** message using the **AES-CMAC** algorithm with the **AppKey**. 
+
+- **cmac** = **aes128_cmac**(**AppKey**, MHDR | AppNonce | NetID | DevAddr | DLSettings | RxDelay | CFList)
+- **MIC** = cmac[0..3]
```
```suggestion
- `cmac = aes128_cmac(AppKey, MHDR | AppNonce | NetID | DevAddr | DLSettings | RxDelay | CFList)`
- `MIC = cmac[0..3]`
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- **MIC** = cmac[0..3]
+ 
+The calculated **MIC** is then added to the **Join-accept** message itself.
+Then the **payload** of the **Join-accept** message is encrypted using **AES decrypt** operation in **ECB** (Electronic Code Book) mode with the **AppKey** so that the end device can use **AES encrypt** operation to **decrypt** the message.
```
```suggestion
The calculated MIC is then added to the Join-accept message.

The payload of the Join-accept message is encrypted using **AES decrypt** operation in **ECB** (Electronic Code Book) mode with the AppKey, so that the end device can use the **AES encrypt** operation to decrypt the message:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+The calculated **MIC** is then added to the **Join-accept** message itself.
+Then the **payload** of the **Join-accept** message is encrypted using **AES decrypt** operation in **ECB** (Electronic Code Book) mode with the **AppKey** so that the end device can use **AES encrypt** operation to **decrypt** the message.
+
+- **aes128_decrypt**(**AppKey**, AppNonce | NetID | DevAddr | DLSettings | RxDelay | CFList | MIC)
```
```suggestion
- `aes128_decrypt(AppKey, AppNonce | NetID | DevAddr | DLSettings | RxDelay | CFList | MIC)`
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+- **aes128_decrypt**(**AppKey**, AppNonce | NetID | DevAddr | DLSettings | RxDelay | CFList | MIC)
+
+Finally the **Network Server** sends the **Join-accept** message to the end device.
```
```suggestion
Finally the Network Server sends the Join-accept message to the end device.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+Finally the **Network Server** sends the **Join-accept** message to the end device.
+
+After the end device receives the Join-accept message it derives two session keys locally, **NwkSKey** and **AppSKey** using AES encrypt.
```
```suggestion
After the end device receives the Join-accept message, it derives two session keys locally, **NwkSKey** and **AppSKey** using AES encrypt:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+After the end device receives the Join-accept message it derives two session keys locally, **NwkSKey** and **AppSKey** using AES encrypt.
+
+- **NwkSKey** = **aes128_encrypt**(**AppKey**, 0x01 | AppNonce | NetID | DevNonce | pad16)
+- **AppSKey** = **aes128_encrypt**(**AppKey**, 0x02 | AppNonce | NetID | DevNonce | pad16)
```
```suggestion
- `NwkSKey = aes128_encrypt(AppKey, 0x01 | AppNonce | NetID | DevNonce | pad16)`
- `AppSKey = aes128_encrypt(AppKey, 0x02 | AppNonce | NetID | DevNonce | pad16)`
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- **NwkSKey** = **aes128_encrypt**(**AppKey**, 0x01 | AppNonce | NetID | DevNonce | pad16)
+- **AppSKey** = **aes128_encrypt**(**AppKey**, 0x02 | AppNonce | NetID | DevNonce | pad16)
+ 
+Similarly the Network Server derives session keys based on the root key and fields in the Join-request and Join-accept messages using AES encrypt. 
```
```suggestion
Similarly, the Network Server derives session keys based on the root key and fields in the Join-request and Join-accept messages using AES encrypt:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+Similarly the Network Server derives session keys based on the root key and fields in the Join-request and Join-accept messages using AES encrypt. 
+
+- **NwkSKey** = **aes128_encrypt**(**AppKey**, 0x01 | AppNonce | NetID | DevNonce | pad16)
+- **AppSKey** = **aes128_encrypt**(**AppKey**, 0x02 | AppNonce | NetID | DevNonce | pad16)
```
```suggestion
- `NwkSKey = aes128_encrypt(AppKey, 0x01 | AppNonce | NetID | DevNonce | pad16)`
- `AppSKey = aes128_encrypt(AppKey**, 0x02 | AppNonce | NetID | DevNonce | pad16)`
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+Finally, the **Network Server** keeps the **NwkSkey** and shares the **AppSKey** with the **Application Server**. 
+- The **AppSKey** is not visible to the **Network Server**.
+- The **NwkSKey** is not visible to the **Application Server**.
```
```suggestion
Finally, the Network Server keeps the NwkSkey and shares the AppSKey with the Application Server. Note that the AppSKey is not visible to the Network Server, and that the NwkSKey is not visible to the Application Server.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+- **aes128_decrypt**(**NwkKey** or **JSEncKey**, JoinNonce | NetID | DevAddr | DLSettings | RxDelay | CFList | MIC)
+ 
+After the end device receives the Join-accept message it derives session keys locally. 
```
Since this is similar to the section above, please apply the changes required there to this section too

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+- **FNwkSIntKey** = **aes128_encrypt**(**NwkKey**, 0x01 | JoinNonce | JoinEUI | DevNonce | pad16 )
+- **SNwkSIntKey** = **aes128_encrypt**(**NwkKey**, 0x03 | JoinNonce | JoinEUI | DevNonce | pad16)
+- **NwkSEncKey** = **aes128_encrypt**(**NwkKey**, 0x04 | JoinNonce | JoinEUI | DevNonce | pad16)
```
```suggestion
The following network session keys are derived using the NwkKey:

- `FNwkSIntKey = aes128_encrypt(NwkKey, 0x01 | JoinNonce | JoinEUI | DevNonce | pad16 )`
- `SNwkSIntKey = aes128_encrypt(NwkKey, 0x03 | JoinNonce | JoinEUI | DevNonce | pad16)`
- `NwkSEncKey = aes128_encrypt(NwkKey, 0x04 | JoinNonce | JoinEUI | DevNonce | pad16)`
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+The **AppSKey** is derived using the **AppKey**.
+
+- **AppSKey** = **aes128_encrypt**(**AppKey**, 0x02 | JoinNonce | JoinEUI | DevNonce | pad16)
```
```suggestion
The AppSKey is derived using the AppKey:

- `AppSKey = aes128_encrypt(AppKey**, 0x02 | JoinNonce | JoinEUI | DevNonce | pad16)`
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+The **OptNeg** field (can be found in the **DLSettings** field) is used to indicate whether the network supports LoRaWAN v1.0.x (unset) or v1.1 and later (set). 
+
+If the **OptNeg is UNSET**, the device **reverts** to the **LoRaWAN v1.0** and the **session keys** are derived from the **NwkKey** as follow:
```
```suggestion
If the OptNeg is unset, the device reverts to the LoRaWAN v1.0 and the session keys are derived from the NwkKey as follow:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+- The **AppSKey** is not visible to the **Network Server**.
+- The **FNwkSIntKey, SNwkSIntKey, NwkSEncKey** are not visible to the **Application Server**.
+ 
+Since both the **end device** and the **network** have session keys, from now all the messages will be **encrypted** using the **AppSKey**, and the **integrity** will be **protected** with the NwkSKey.
```
Same as above, please make sure that you don't put in **bold** things that are already introduced, put these "formulas" under `quotes` and put : before enumeration.

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+Every uplink and downlink data messages (both unconfirmed and confirmed) carry either **MAC commands, application data**, or both **application data and MAC commands together**. The **MACPayload** of the data message contains a field called **FRMPayload** which is capable of carrying either a **sequence of MAC commands** (between the end device and the Network Server) or **application data** (bet ...
+ 
+{{< note "LoRaWAN uses a minor variation of the CCM, which is CCM*." />}}
```
```suggestion
## Encrypting the FRMPayload Field

Every uplink and downlink data message (both unconfirmed and confirmed) carries MAC commands, application data, or both. The MACPayload of the data message contains a field called FRMPayload which is capable of carrying either a sequence of MAC commands (between the end device and the Network Server) or application data (between the end device and the Application Server). The FRMPayload field is encrypted using AES encrypt operation in CCM mode (counter with cipher block chaining message authentication code; counter with CBC-MAC). The encryption key used to encrypt the FRMPayload field depends on what it contains. Note that LoRaWAN uses a minor variation of the CCM, named CCM*.
```

nejraselimovic replied:
This is a bit misguiding. You first say that the data message can carry either MAC commands, app data or both, then after that you say that FRMPayload can carry either a sequence of MAC commands or app data, but not both. Please elaborate this a bit better.

pradeeka7 replied:
I agree with you. So I removed the first sentence "_Every uplink and downlink data message (both unconfirmed and confirmed) carries MAC commands, application data, or both._". 
I think it is now clear.

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+{{< figure src="frmpayload-encryption.png" alt="" >}}
+
+## Encrypting FOpts field - MAC Commands
```
```suggestion
## Encrypting FOpts field
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+The **FOpts** field is **encrypted** using **AES encrypt** operation in CCM* mode. 
+ 
+The following keys are used to **encrypt** the **FOpts** field in the MACPayload in different LoRaWAN versions for both uplink and downlink data messages.
```
Remove the bold style

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+{{< figure src="fopts-encryption.png" alt="" >}}
+
+## Computing Message Integrity Code
```
```suggestion
## Computing MIC
```

nejraselimovic replied:
You introduced the MIC abbreviation for a reason - to not have to write Message Integrity Code in this section anymore

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+## Computing Message Integrity Code
+
+After encrypting the payload the MIC is calculated. The Message Integrity Code is computed for each packet (entire PHYPayload) using the AES-CMAC (AES Cipher-based Message Authentication Code).
```
```suggestion
The MIC is calculated after encrypting the payload. The MIC is computed for each packet (entire PHYPayload) using the AES-CMAC.
```

nejraselimovic replied:
Same as for MIC, I think you introduced a full name for CMAC in the text above

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+{{< figure src="mic-compute-keys.png" alt="" >}}
+
+When a **LoRaWAN 1.1 device** is provisioned with a **LoRaWAN 1.0.x Network Server**, the **MIC** of the uplink and downlink **data messages** are computed with the **FNwkSIntKey**.
```
Remove bold style

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+When a **LoRaWAN 1.1 device** is provisioned with a **LoRaWAN 1.0.x Network Server**, the **MIC** of the uplink and downlink **data messages** are computed with the **FNwkSIntKey**.
+ 
+## End-to-end security
```
```suggestion
## End-to-end Security
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+When a **LoRaWAN 1.1 device** is provisioned with a **LoRaWAN 1.0.x Network Server**, the **MIC** of the uplink and downlink **data messages** are computed with the **FNwkSIntKey**.
+ 
+## End-to-end security
+LoRaWAN provides end-to-end encryption for application payloads exchanging between the end devices and the application servers. Application payload refers to the data contained in the FRMPayload field. The Network Server also verifies the DevAddr (device address) for authenticating the received messages.
```
```suggestion

LoRaWAN provides end-to-end encryption for application payload exchange between end devices and the Application Server. Application payload refers to the data contained in the FRMPayload field. The Network Server also verifies the DevAddr for authenticating received messages.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+LoRaWAN provides end-to-end encryption for application payloads exchanging between the end devices and the application servers. Application payload refers to the data contained in the FRMPayload field. The Network Server also verifies the DevAddr (device address) for authenticating the received messages.
+ 
+## Replay Protection
+Replay protection ensures that the receiver does not accept an already received frame.
```
```suggestion
## Replay Protection

Replay protection ensures that the receiver does not accept a frame that has already been received.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+In **LoRaWAN v1.1**, DevNonce is a counter starting at 0. It is used to keep track of the Join-request messages. The DevNonce value is incremented with every Join-request. The Network Server keeps track of the last DevNonce value used by the end device and ignores Join-requests if DevNonce is not incremented.
+ 
+Similarly, the replay protection for the ReJoin-request messages is implemented with the RJcount0 (for ReJoin-request Type 0 or 2 messages) or RJcount1 (for ReJoin-request Type 1 messages).
```
```suggestion
### Join-request Messages

The **replay protection** of the **Join-request** messages is protected using **DevNonce**. The DevNonce field is used when computing the MIC.

In **LoRaWAN 1.0.x**, DevNonce is a random value. It is used to keep track of the Join-request messages. The network server keeps track of a certain number of DevNonce values for each end device. It ignores any Join-requests that include a previously recorded DevNonce value.

In **LoRaWAN v1.1**, DevNonce is a counter starting at 0. It is used to keep track of the Join-request messages. The DevNonce value is incremented with every Join-request. The Network Server keeps track of the last DevNonce value used by the end device and ignores Join-requests if DevNonce is not incremented.
 
Similarly, the replay protection for the Rejoin-request messages is implemented with the RJcount0 (for Rejoin-request type 0 or 2 messages) or RJcount1 (for Rejoin-request type 1 messages).
```

nejraselimovic replied:
Drop the bold style. Also, I think you were using Rejoin-request in the text above, so please be consistent and don't use ReJoin-request here

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+In **LoRaWAN 1.0.x** two frame counters are used to keep track of uplink and downlink messages - **FCntUP** and **FCntDown**. These frame counters are stored in both the end device and the Network Server. Frame counters are initially set to 0 and incremented by the value 1 with each transmission.
+ 
+- **FCntUp** - keeps track of the number of data messages sent from the end device to the Network Server.
+- **FCntDown** - keeps track of the number of data messages sent from the Network Server to the end device.
```
```suggestion
In **LoRaWAN 1.0.x** two frame counters are used to keep track of uplink and downlink messages:
 
- **FCntUp** - keeps track of the number of data messages sent from the end device to the Network Server
- **FCntDown** - keeps track of the number of data messages sent from the Network Server to the end device

These frame counters are stored in both the end device and the Network Server. Frame counters are initially set to 0 and incremented by the value 1 with each transmission.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+If either the device or the network receives a message with a frame counter that is lower than the last one, the message is discarded.
+
+In **LoRaWAN 1.1** each end device maintains three frame counters to keep track of uplink and downlink data messages. 
```
```suggestion
In **LoRaWAN 1.1** each end device maintains three frame counters to keep track of uplink and downlink data messages:
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+ 
+Similarly, the replay protection for the ReJoin-request messages is implemented with the RJcount0 (for ReJoin-request Type 0 or 2 messages) or RJcount1 (for ReJoin-request Type 1 messages).
+
+### Data Messages
```
Drop the bold for everything that's not mentioned for the first time

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+## Eavesdropping
+
+Since LoRaWAN provides two layers of security, the MAC payload is encrypted between the end device and the Network Server whereas the application payload is encrypted between the end device and the Application Server. This ensures only the authorised parties (end device, network server, application server) that hold the security keys (network session key/s and the application session key) can dec ...
```
```suggestion
Since LoRaWAN provides two layers of security, the MAC payload is encrypted between the end device and the Network Server whereas the application payload is encrypted between the end device and the Application Server. This ensures only the authorised parties (end device, Network Server, Application Server) that hold the security keys (NwkSKey and AppSKey) can decrypt the payload and read the plain text content. 
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+Since LoRaWAN provides two layers of security, the MAC payload is encrypted between the end device and the Network Server whereas the application payload is encrypted between the end device and the Application Server. This ensures only the authorised parties (end device, network server, application server) that hold the security keys (network session key/s and the application session key) can dec ...
+
+## Spoofing
+The MACPayload in the data message is origin-authenticated and integrity protected with the MIC. This ensures only the authorised parties (end device, network server, application server) that hold the security keys (network session key/s and the application session key) can generate valid data frames.
```
```suggestion
## Spoofing

The MACPayload in the data message is origin-authenticated and integrity protected with the MIC. This ensures only the authorised parties (end device, Network Server, Application Server) that hold the security keys (NwkSKey and AppSKey) can generate valid data frames.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+The MACPayload in the data message is origin-authenticated and integrity protected with the MIC. This ensures only the authorised parties (end device, network server, application server) that hold the security keys (network session key/s and the application session key) can generate valid data frames.
+
+## Securing communication between network elements
+LoRaWAN uses HTTPS and VPN for securing communication among network elements (backend interfaces) such as the Network Server, Application Servers, and the Join Server.
```
```suggestion
## Securing communication between network elements

LoRaWAN uses HTTPS and VPN for securing communication among network elements (backend interfaces) such as the Network Server, Application Servers, and the Join Server.
```

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+## Secure Elements
+
+The secure element is a small chip, for example, Microchip’s hardware secure element (ATECC608B-TNGLORA), that is used to securely store root keys and for performing cryptographic operations. The root keys are securely stored in the secure elements during the manufacturing process. Once stored it is extremely difficult to read the root keys from the chip. Simultaneously the matching root key/s fo ...
```
```suggestion
The secure element is a small chip, for example Microchip’s hardware secure element (ATECC608B-TNGLORA), that is used to securely store root keys and for performing cryptographic operations. The root keys are securely stored in the secure elements during the manufacturing process. Once stored, it is extremely difficult to read the root keys from the chip. Simultaneously, the matching root key/s for each secure element are provisioned on the [Global Join Server](https://www.thethingsindustries.com/security/). 
```

nejraselimovic replied:
Link this https://www.thethingsindustries.com/docs/devices/atecc608a/

nejraselimovic commented on doc/content/lorawan/the-things-certified-security/_index.md:
```diff
@@ -0,0 +1,289 @@
+
+Device manufacturers can easily integrate these secure element chips with their LoRaWAN end devices. During the join procedure, the secure element generates an identical pair of session keys from the root key/s, the same way the Join Server does it.
+
+The Global Join Server allows the end-user to transfer these root keys anytime to a third-party Join Server. Once transferred the end device generates new root keys by using the rekeying procedure.
```
```suggestion
The Global Join Server allows the end-user to transfer these root keys anytime to a third-party Join Server. Once transferred, the end device generates new root keys using the rekeying procedure.
```

nejraselimovic reviewed (changes requested):
Please make sure to apply all requested changes to the whole doc, because there are high chances that I didn't mark the stuff that is repeated more than 2 times. Please remember that we also have a documentation page here https://www.thethingsindustries.com/docs/ so you can link some stuff to it, for example why OTAA is better than ABP, hardware secure elements, etc. You can tag me for another round of review once you apply the requested changes

jpmeijers commented:
Small comment @pradeeka7 - Do you mind making all the changes first before you push. Everyone subscribed to this repo gets a notification for every new push.

pradeeka7 commented:
> Small comment @pradeeka7 - Do you mind making all the changes first before you push. Everyone subscribed to this repo gets a notification for every new push.

@jpmeijers Thanks for letting me know and will do accordingly.

pradeeka7 commented:
@jpmeijers - I haven't touched the 'Resolve conversation' buttons because I am not sure if it sends notifications to everyone.

@nejraselimovic - Thanks for reviewing. I have addressed all the issues. Please read my comments (a very few) too.
